### PR TITLE
add formatting style

### PR DESCRIPTION
**Pull request by skrax**
- add _.clang-format_ based on google
- add check-formatting workflow
- reformat all files to new format

#39

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  AfterClass:      false
+  AfterControlStatement: false
+  AfterEnum:       false
+  AfterFunction:   false
+  AfterNamespace:  false
+  AfterObjCDeclaration: false
+  AfterStruct:     false
+  AfterUnion:      false
+  AfterExternBlock: false
+  BeforeCatch:     false
```
Comment (apodrugin) [inline .clang-format:38]: Do you mind if we change BeforeCatch, BeforeElse to true and also set BeforeWhile to true?

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Attach
+BreakBeforeInheritanceComma: false
+BreakInheritanceList: BeforeColon
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+BreakConstructorInitializers: BeforeColon
+BreakAfterJavaFieldAnnotations: false
+BreakStringLiterals: true
+ColumnLimit:     80
```
Comment (apodrugin) [inline .clang-format:53]: Can we change this to 120 at least? Nowadays everybody has widescreen monitors... 😄

File: .clang-format [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
+AlignEscapedNewlines: Left
+AlignOperands:   true
+AlignTrailingComments: true
+AllowAllArgumentsOnNextLine: true
+AllowAllConstructorInitializersOnNextLine: true
+AllowAllParametersOfDeclarationOnNextLine: true
+AllowShortBlocksOnASingleLine: Never
+AllowShortCaseLabelsOnASingleLine: false
```
Comment (apodrugin) [inline .clang-format:16]: I propose to forbid single line blocks in all cases. This makes code more consistent.

File: ubuntu/check-formatting.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+FILES=$(find src -regex ".*\.\(hpp\|cpp\|cc\|c\|h\)$")
```
Comment (apodrugin) [inline ubuntu/check-formatting.sh:2]: It looks like this script can also be run on MacOS without problems. Probably it is better to store such scripts to some common folder (for example we can name it 'unix').

**Review by apodrugin — CHANGES_REQUESTED** [review]
Good job 👍. I have few code style suggestions. Also it would be nice to add braces to single line if statements. I've made small research and it seems clang format does not handle this (according to https://stackoverflow.com/questions/26111162/can-clang-format-add-braces-to-single-line-if-statements-etc) and the clang-tidy is the only solution https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html

File: .clang-format (hunk 2 above)
Comment (skrax) [reply .clang-format:53]: I disagree, since often I need to fit two files vertically.

File: .clang-format (hunk 3 above)
Comment (skrax) [reply .clang-format:16]: Will change.

File: .clang-format (hunk 2 above)
Comment (apodrugin) [reply .clang-format:53]: Editor you use does not perform lines wrapping automatically?

File: .clang-format (hunk 1 above)
Comment (skrax) [reply .clang-format:38]: Will change.

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    SortPriority:    0
+  - Regex:           '.*'
+    Priority:        3
+    SortPriority:    0
+IncludeIsMainRegex: '([-_](test|unittest))?$'
+IncludeIsMainSourceRegex: ''
+IndentCaseLabels: true
+IndentGotoLabels: true
+IndentPPDirectives: None
+IndentWidth:     2
```
[the remark this answers is not in the document]
Comment (skrax) [reply .clang-format:88]: I think 2 is plenty, but could live with 4 if the majority is in favor.

File: .clang-format (hunk 2 above)
Comment (skrax) [reply .clang-format:53]: It can, but I prefer not to with code editors.

Comment (apodrugin) [reply .clang-format:53]: In Xcode it is possible to do crazy things like this 
Screen Shot 2021-03-15 at 12 09 07

File: ubuntu/check-formatting.sh (hunk 4 above)
Comment (skrax) [reply ubuntu/check-formatting.sh:2]: Sure.

File: .clang-format (hunk 2 above)
Comment (skrax) [reply .clang-format:53]: Well, not everybody is using Xcode.

File: .clang-format [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  AfterEnum:       false
+  AfterFunction:   false
+  AfterNamespace:  false
+  AfterObjCDeclaration: false
+  AfterStruct:     false
+  AfterUnion:      false
+  AfterExternBlock: false
+  BeforeCatch:     true
+  BeforeElse:      true
+  IndentBraces:    false
```
Comment (apodrugin) [inline .clang-format:40]: What about BeforeWhile? Is the default value 'true'?

File: .clang-format [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  SplitEmptyFunction: true
+  SplitEmptyRecord: true
+  SplitEmptyNamespace: true
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Attach
```
Comment (apodrugin) [inline .clang-format:45]: It looks like that BraceWrapping is ignored if BreakBeforeBraces is not set to 'Custom' (see BraceWrapping  section in https://clang.llvm.org/docs/ClangFormatStyleOptions.html)

File: .clang-format (hunk 5 above)
[the remark this answers is not in the document]
Comment (apodrugin) [reply .clang-format:88]: It looks like that nobody cares about it... 😞

**Review by apodrugin — CHANGES_REQUESTED** [review]

File: .clang-format (hunk 6 above)
Comment (skrax) [reply .clang-format:40]: BeforeWhile is not available with ClangFormat 10.0 (Ubuntu 20.04)
We can build a newer version, not sure if we can submodule it. I will look into it.

Comment (apodrugin) [reply .clang-format:40]: I think we can give it up if it takes a lot of effort.

Comment (skrax) [reply .clang-format:40]: It's not alot of effort, but I think it's easier to just go with Ubuntu20. It's really just the config file, which makes a difference.

**Pull request merged**